docs/adr/0007-modelling-history.md

```diff
@@ -0,0 +1,272 @@
+ids/lookup data needed for asset manager. It is hoped that this model can be
+shared with file attachments. Image objects that reference the same file on
+asset manager can share a state.
+
+The following diagram represents the objects of these classes to give an
+example of how a document could look at a point in a life cycle:
+
+![Object diagram](0007/object-diagram.png)
+
+## Status
+
+Proposed
+
+## Consequences
+
+This modelling should provide a basis to provide the richer history
+functionality that can enable meeting and exceeding the features offered by
+Whitehall and Mainstream Publisher.
+
+It does however come with a number of concerns that are highlighted below.
+
+### Mo' models, mo' problems
+
+With an increase in the number of models there is more for a developer to
+understand and a higher barrier for entry. For every action there are either
+more database queries or a more complicated single query to look up information.
+
+Some ideas to alleviate this are:
+
+- the use of [delegate][] with ActiveRecord to make it simpler to perform
+  methods involved in the modelling;
+- use of [scopes][] / find methods to alleviate join knowledge;
+- potentially for a [materialized view][] (or
+  [search index][elasticsearch-rails]) to have current and live data in a
+  single table.
+
+### There will be data repetition
+
+By storing a variety of data on a revision directly this data will then be
+replicated on the next revision which may be inefficient for data usage. This
+seems a natural consequence of storing all revisions of a document without
+going down a complicated route of difference storing. However an idea to
+alleviate this is discussed in
+[further options](#store-less-data-on-a-revision).
+
+### We're not considering removing the past
+
+This data model is built around the expectation that the history of documents
+published to GOV.UK is something that, at this current point in time, we want to
+preserve. Over time this may lead to a large amount of data usage and need
+contemplating. However the evidence of 6 years of GOV.UK does not suggest that
+the data usage is currently a high concern thus this design considers this a
+future problem that may or may not have to be solved.
+
+## Further options being considered
+
+This section outlines some ideas of subtle or supporting changes to this
+modelling that is being considered
+
+### Store less data on a revision
+
+This proposal focuses on copying a revision each time an aspect of it changes,
+whether that is content, tags, or images. This would mean that if we had a 1000
+word document those 1000 words would be repeated in the database for each
+change to images of that document. A way to resolve this would be for a revision
+to be used to just track foreign keys of different data of a document
+
+![Diagram of expanded revision concept](0007/revision-expanded-diagram.png)
+
+Doing this could also give us a more granular level to know the last version
```

Comment:
Yup, I can add something - wasn't meant to sound like an embedded assumption

What I'm thinking is if you submitted a change to Title one of the parameters could be the content_change_id and then at the point of handling the request you could check whether the revision still had content at that id - if not you know it has changed since you started your work. Does that make sense?